sky/packages/sky/lib/widgets/drag_target.dart

```diff
@@ -0,0 +1,119 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:collection';
+
+import 'package:sky/base/hit_test.dart';
+import 'package:sky/rendering/sky_binding.dart';
+import 'package:sky/widgets/basic.dart';
+import 'package:sky/widgets/framework.dart';
+
+typedef bool DragTargetWillAccept(dynamic data);
+typedef void DragTargetAccept(dynamic data);
+typedef Widget DragTargetBuilder(List acceptedData, List rejectedData);
+
+class DragTarget extends StatefulComponent {
+  DragTarget({
+    Key key,
+    this.builder,
+    this.onWillAccept,
+    this.onAccept
+  }) : super(key: key);
+
+  DragTargetBuilder builder;
+  DragTargetWillAccept onWillAccept;
+  DragTargetAccept onAccept;
+
+  final List _acceptedData = new List();
+  final List _rejectedData = new List();
+
+  void syncFields(DragTarget source) {
+    builder = source.builder;
+    onWillAccept = source.onWillAccept;
+    onAccept = source.onAccept;
+  }
+
+  bool didEnter(dynamic data) {
+    assert(!_acceptedData.contains(data));
+    assert(!_rejectedData.contains(data));
+    if (onWillAccept != null && onWillAccept(data)) {
+      setState(() {
+        _acceptedData.add(data);
+      });
+      return true;
+    }
+    _rejectedData.add(data);
+    return false;
+  }
+
+  void didLeave(dynamic data) {
+    assert(_acceptedData.contains(data) || _rejectedData.contains(data));
+    setState(() {
+      _acceptedData.remove(data);
+      _rejectedData.remove(data);
+    });
+  }
+
+  void didDrop(dynamic data) {
+    assert(_acceptedData.contains(data));
+    setState(() {
+      _acceptedData.remove(data);
+    });
+    if (onAccept != null)
+      onAccept(data);
+  }
+
+  Widget build() {
+    return builder(new UnmodifiableListView(_acceptedData),
+                   new UnmodifiableListView(_rejectedData));
+  }
+}
+
+class DragController {
+  DragController(this.data);
+
+  final dynamic data;
+
+  DragTarget _focus;
```

Comment:
call this something other than "focus" to avoid confusion with keyboard focus in focus.dart